### ISSUE TITLE
[Bounty] Unlock the imagination of online publication directly on Twitter with Unlock Protocol

### ISSUE THREAD

**Issue by yisiliu** [DimensionDev/Maskbook#2363]
## Catalogue

- Platform
  - [x] Gitcoin
  - [ ] Hackerlink
  - [ ] Mask
- Experience
  - [ ] Beginner
  - [x] Intermediate
  - [ ] Advanced
- Category
  - [ ] Mask Core Feature
  - [x] Mask Plugin
  - [ ] Security/Cryptography Feature
  - [ ] Documentation
  - [ ] Design
  - [ ] Other
    - [ ] /_List it here_/
- Time Commitment
  - [x] 4 Weeks
- Bounty Size
  - [x] 6000 USD

## Part 1: Problem Description

Mask offers a browser extension which lets users create encrypted tweets where the content of the tweet can only be “viewed” by other Twitter user if they use the Maskbook browser extension. These tweets can include text, but also images.

Unlock is a protocol for memberships which lets creators deploy a lock and sell access to this lock using the Ethereum blockchain. Locks can be added to websites, or applications in order to monetize them.

We are proposing an integration between Unlock and Mask where creators will be able to post encrypted Tweets and put them “behind” a lock, so that only members of the lock are able to view their content. If viewers are not members, then, Mask would offer the ability to Unlock these memberships. In that scenario, Mask is a referrer and should earn UDT.

For a more detailed description, you can refer to [this](https://docs.google.com/document/d/18bG3NAbfFZ7ofUEtq1Rdjs29XM1HveQ9htZ7rvC8JGo/edit?usp=sharing).

## Part 2: Resources

https://unlock-protocol.com/
https://github.com/unlock-protocol/unlock

## Part 3: Contact and Support

[Mask Discord Server](https://discord.gg/4Xccj7a75Q)
@yisiliu @Jack-Works -> Mask
@julien51 -> Unlock

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 2000.0 DAI (2000.0 USD @ $1.0/DAI)  attached to it.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701).
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://discord.com/invite/83BK5z55yB/'>Gitcoin's Discord</a>
* $1,060,306.36 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

__Work has been started__.


These users each claimed they can complete the work by 265 years, 9 months from now.
Please review their action plans below:


**1) [iamzubin](https://gitcoin.co/iamzubin) has been approved to start work.**

Hey, I read all the requirements and I think I will be able to do the job in a few week, The google doc is very helpful and the code is well documented which would be really helpful in for the development of the feature, We can discuss more about the project on discord.

Thanks, looking forward for working on this project
**2) [koosy79](https://gitcoin.co/koosy79) has applied to start work _(Funders only: [approve worker](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?mutate_worker_action=approve&worker=koosy79) | [reject worker](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?mutate_worker_action=reject&worker=koosy79))_.**

Something like the limitless possibilities protocol or take out the possibilities
Or something absolutely so amazingly pompous that people will have to investigate

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701).

**Comment by gitcoinbot** [comment]
@iamzubin Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=1) | [3 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=3) | [5 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=5) | [10 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=10) | [100 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=100)

**Comment by gitcoinbot** [comment]
@iamzubin Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=1) | [3 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=3) | [5 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=5) | [10 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=10) | [100 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=100)

**Comment by gitcoinbot** [comment]
@iamzubin Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=1) | [3 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=3) | [5 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=5) | [10 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=10) | [100 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=100)

**Comment by gitcoinbot** [comment]
@iamzubin Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=1) | [3 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=3) | [5 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=5) | [10 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=10) | [100 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=100)

**Comment by iamzubin** [comment]
bump

**Comment by gitcoinbot** [comment]
@iamzubin Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=1) | [3 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=3) | [5 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=5) | [10 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=10) | [100 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=100)

**Comment by iamzubin** [comment]
bump

**Comment by gitcoinbot** [comment]
@iamzubin Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=1) | [3 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=3) | [5 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=5) | [10 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=10) | [100 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=100)

**Comment by iamzubin** [comment]
bump

**Comment by iamzubin** [comment]
bump

**Comment by Jack-Works** [comment]
So let me try to describe how this might be work:

1. Creator:
        1. Let `content` be the content to be encrypted (input by the creator).
        2. Let `K` be a pair of randomly generated EC keys.
        3. Let `AES` be a random AES key.
        4. Let `encrypted` be `encrypt(content, AES)`
        5. Let `hint` be `encrypt([K.priv, AES], localKey)` (Note: `localKey` is an AES key in Mask. It is used to help the author to store some really private information. Question: Or the self public key?).
        6. Let `id` be a new UUID.
        7. Set Metadata of the post to contains `id`, `encrypted`, `K.public`, `hint`, `pay_address`.
        8. Listen to the channel `unlock-ping-$(cut_hash(id))`
2. Customer:
        1. Get `id`, `encrypted`, `K.public`, `hint`, `pay_address`.
        2. Once paid to the `pay_address`, continue.
        3. Let `K2` be a pair of randomly generated EC keys.
        4. Let `req` be `encrypt([ customer_wallet_addr, K2.public ], K.public)`
        5. Send `req` to the channel `unlock-ping-$(cut_hash(id))`
        6. Listen to the channel `unlock-pong-$(cut_hash(K2.public))`
3. Creator: (receiving a message from the channel `unlock-ping-$(cut_hash(id))` above)
        1. Let [`addr`, `K2.public`] be decrypted message with `K.private`, if failed, ignore the message.
        2. Verify if `addr` has paid on the blockchain, if not, ignore the message. (Maybe we should wait for the block confirmation?)
        3. Let `key` be `encrypt(AES, pub)`.
        4. Send `key` to  the channel `unlock-pong-$(cut_hash(K2.public))`
4. Customer: (receiving a message from the channel `unlock-pong-$(cut_hash(K2.public))`)
        1. Let `key` be decrypted message with `K2.private`, if failed, ignore the message.
        2. Try to decrypt the unlock message

**Comment by gitcoinbot** [comment]
@iamzubin Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=1) | [3 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=3) | [5 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=5) | [10 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=10) | [100 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=100)

**Comment by gitcoinbot** [comment]
@iamzubin Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=1) | [3 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=3) | [5 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=5) | [10 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=10) | [100 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=100)

**Comment by iamzubin** [comment]
Bump

**Comment by thisispalash** [comment]
Also, the updated bounty quote as per the new requirements with a [rough] time breakdown and payment milestones, as discussed with @iamzubin.

## updated bounty quote [160h &Rightarrow; 6000 USD]

### ideation `10 March` &Rightarrow; 500 USD
 - create protocol for key discovery :: 16h

### server `20 March` &Rightarrow; 2000 USD
 - Create nodes for message broadcasting :: 16h
 - Sync broadcasting nodes :: 32h

### client `28 March` &Rightarrow; 2000 USD
 - Key discovery :: 32h
 - Verify transaction :: 32h
 
### rendering `31 March` &Rightarrow; 1500 USD
 - Connect wallets (unlock, payment) :: 16h
 - Create post / select recepients :: 16h

#### Payment Milestone
 - `20 March` &Rightarrow; 2500 USD :: ideation and server
 - `31 March` &Rightarrow; 3500 USD :: client and rendering

**Comment by yisiliu** [comment]
@iamzubin @thisispalash the bounty size is updated. Good luck guys! If you need our help to expedite your development, please let us know. Our team would like to help!

**Comment by gitcoinbot** [comment]
@iamzubin Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=1) | [3 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=3) | [5 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=5) | [10 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=10) | [100 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=100)

**Comment by iamzubin** [comment]
Bump

**Comment by gitcoinbot** [comment]
@iamzubin Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=1) | [3 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=3) | [5 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=5) | [10 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=10) | [100 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=100)

**Comment by iamzubin** [comment]
bump

**Comment by gitcoinbot** [comment]
@iamzubin Hello from Gitcoin Core - are you still working on this issue? Please submit a WIP PR or comment back within the next 3 days or you will be removed from this ticket and it will be returned to an ‘Open’ status. Please let us know if you have questions!
* [x] reminder (3 days)
* [ ] escalation to mods (6 days)

Funders only: Snooze warnings for [1 day](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=1) | [3 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=3) | [5 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=5) | [10 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=10) | [100 days](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701?snooze=100)

**Comment by iamzubin** [comment]
Bump :3

**Comment by yisiliu** [comment]
@iamzubin hey do you guys have any updates?

**Comment by Tedko** [comment]
also cc @thisispalash

**Comment by iamzubin** [comment]
Hey, 
yeah there's been some delay.
We're working on it.

**Comment by thisispalash** [comment]
Hi,

I apologise for the lack in communication, my schoolwork piled up significantly. We should be on track to finish up by next weekend (ie, by Apr 4) instead of Mar 31. 

Presently we are working in parallel with @iamzubin handling the client side features and me handling the web socket server. For the server specifically, I need to implement a Trie to store the UUIDs of the channels (and addresses subscribed to the channel as a list). Post that, just a simple server sync would be left (either via redis, or manually).

As for the client side, @iamzubin has finished the verification of transactions. He'd be doing the logic for displaying the tweet / unlock button next. Post that, the key management by talking to the server.

After integrating everything, we'll write the tests (which is probably bad practice but what the hell) to submit the PR. We may need assistance in writing said tests.

I apologise again for the delay in response.

We mostly will not be working this weekend due to Holi (for @iamzubin) and me having two exams early next week. Hence, the shift in deadline.

**Comment by yisiliu** [comment]
@thisispalash thanks for syncing with us and you guys are totally fine but please do let us know if you need any help since we would like to prioritize this feature in our plan. @julien51 what do you think?

**Comment by julien51** [comment]
Works for me! Please do let me know if you get stuck...

**Comment by yisiliu** [comment]
Hey @thisispalash @iamzubin any update? Please create a pr so that we can start reviewing it ASAP

**Comment by thisispalash** [comment]
Hey, so we have been working in a separate repository. I've made it public and you can view the [code](https://github.com/thisispalash/bounty-maskbook). I've been on the [server branch](https://github.com/thisispalash/bounty-maskbook/tree/server) while @iamzubin has been working on the frontend on the [unlock branch](https://github.com/thisispalash/bounty-maskbook/tree/unlock).

We still have to write tests and integrate a bunch of things. Sorry about blowing through the weekend deadline, but you can review the code at least. I think this should be done in another week, but I am not certain as there are a lot of firsts for me in this project and my estimation has clearly been off.

**Comment by yisiliu** [comment]
Thanks @thisispalash we will start reviewing it later and see if we can assist you guys there

**Comment by thisispalash** [comment]
Sure that works. Let us know if there are discrepancies you see in conventions or if there are better practices. Off the top of my head, I know some files have mismatched spacing but I am not too sure if that affects anything.
Also, my discord is `k!ller#6839` if you don't want to make this thread super long (already have to scroll quite a bit).

**Comment by Jack-Works** [comment]
please cc me if u need a review on https://github.com/thisispalash/bounty-maskbook

**Comment by yisiliu** [comment]
@iamzubin hey man what's your update on this bounty? I was reading your unlock branch but didn't see much update there.

**Comment by thisispalash** [comment]
@Jack-Works yes please. Specifically within this [folder](https://github.com/thisispalash/bounty-maskbook/tree/server/server-ts) (`server-ts` on the `server` branch).

**Comment by iamzubin** [comment]
> 
> 
> @iamzubin hey man what's your update on this bounty? I was reading your unlock branch but didn't see much update there.

Umm yeah, I've been working on my fork of [Maskbook](https://github.com/iamzubin/Maskbook/tree/released) 
Ran into some issues, resolved now. Will PR soon.

**Comment by iamzubin** [comment]
Hey, @yisiliu , Sorry for the delay, I've been working on my fork of [Maskbook](https://github.com/iamzubin/Maskbook/tree/released).
I'll be pushing and making the PR soon

**Comment by yisiliu** [comment]
@iamzubin thanks! Looking forward to it!

**Comment by Tedko** [comment]
Good job guys @iamzubin @thisispalash looking forward. pls keep us updated

**Comment by yisiliu** [comment]
@thisispalash @iamzubin hey guys I hate disturbing people but can you have us updated please? It's been a while and we really want it to be merged, soon.

**Comment by iamzubin** [comment]
Yeah, apologies for the delayed response on github, I'm being active on the discord group.

here are some screenshots of the progress 
![image](https://user-images.githubusercontent.com/8392856/115462442-4726c400-a248-11eb-9d57-0817698c0e8f.png)
![image](https://user-images.githubusercontent.com/8392856/115462324-26f70500-a248-11eb-957b-160f4e237880.png)

we'll be trying to finish up the project ASAP.

**Comment by iamzubin** [comment]
Hey @Jack-Works  @yisiliu,

I just finished with all the client side logic and UI elements.
Meanwhile @thisispalash  was working on the server for the key sharing.

As you explained in the issue #2520
> It must meet the following requirement:
> 
> * Have none or very low network traffic when there is no message for this client wants.
> * Opt-in Subscribe to certain message types (e.g. `plugin/unlock/key_hash_partition_x`).
> * Privacy concerns, no meaning info leak to it.
> 
> It would be better if it meets the following requirement:
> 
> * Decentralized like matrix (multiple servers) can connect with each other
> * Can store and replay the messages for a certain time in case the client wants to receive the event when it is offline.

We decided to build a web socket client and chose Redis for different server hosts to sync.

Is there any other way way we can write the server so we can also hold the message and so the Decentralised network is not essentially centralised (Redis).

**Comment by iamzubin** [comment]
Oh! By design as discussed in the discord chat
This service could replace the current system in place (gun database). So it should be designed keeping in mind.

**Comment by Jack-Works** [comment]
Hi, @thisispalash I have reviewed the `server-ts` folder in the `server` branch. I have the following review suggestions:

1. `nodemon.json` points to `src/server.ts` but there is no such file. Only `src/server.js` and `src/.server.js`
2. This project should have package-lock file checked in.
3. Please clean up unused files. I don't know which file is the main entry so I guess `src/.server.js` is.
4. I suggest using [msgpack](https://npmjs.org/@msgpack/msgpack) to compress the message, it will make the network traffic smaller.
5. I suggest using the number enum for UserAction, it will make the network traffic smaller.
6. `'use strict';` is no need for a ES Module file.
7. `<reference path='../declarations/node.d.ts' />` there is no such file in the project
8. `declare id : string;` is unused.
9. `let CHANNELS: { [key: string]: Array<string> } = {}` Please use [Map](https://mdn.io/Map) and [Set](https://mdn.io/Set) to represent the subscribers and the channels.
10. What is `Message.identifier`?
11. If you change `Message.payload` to unknown (to allow any JSON object) it will be more efficient to compressed.
12. No need to specify `"node_modules/@types"` in `typeRoots`. That's the default behavior.

The logic seems a bit strange, I need to review the logic again when the code has been cleanuped. Thanks.

**Comment by Jack-Works** [comment]
Oops, I reviewed that in the wrong branch. Let me do that again

**Comment by Jack-Works** [comment]
Hi @thisispalash can you write a README in the repo to explain what are `UserAction` `UserChannel` `Trie` and how those things composed? I'd like to understand this correctly before review this.

**Comment by iamzubin** [comment]
Hey, @thisispalash won't be working on this project anymore, So I would be taking over the server code as well.
This is basically the structure for the request
example for a request for a post with the `cut-hash-id = 123abc`

`{
"identifier" : "123abc",
"action" : "join"
"channel" : "ping"
}`


about trie, it's a data structure @thisispalash decided to use to speed up the lookup for the room identifier
[Wikipedia Trie](https://en.wikipedia.org/wiki/Trie)

**Comment by iamzubin** [comment]
What I'm actually more concerned about is the basic working of the server, as it being a websocket server we don't currently have a way to relay the messages to perform the key exchange in multiple sessions. we decided it should be fine as long as any one user that has the key to unlock the post is online would share the key.

another concern of mine was about the redis database, and the usage of a pub/sub client to sync up multiple websocket server if someone(or later we) want to add new servers. this was the best solution we could come up with, if you have any other suggestions. ( as explained above it matches the "must" requirements but not the "would be better/nice to have" ones)

**Comment by iamzubin** [comment]
Hey, @yisiliu @Tedko I’ve been trying to contact you guys on telegram but @yisiliu hasn’t  been replying since a few weeks and @Tedko stopped replying as well :(

Can I please get a split of the payment as discussed in the issue here 
https://github.com/DimensionDev/Maskbook/issues/2363#issuecomment-794577313

Also @Jack-Works please take a look at the server so I can continue working on implementing the feature

cc @julien51

**Comment by Jack-Works** [comment]
> Also @Jack-Works please take a look at the server so I can continue working on implementing the feature

Yes sure, sorry for the delay

**Comment by Jack-Works** [comment]
@iamzubin Hi can you open a PR in your repo from the server branch to the main/master branch? Therefore I can provide my review suggestions in that place.

**Comment by iamzubin** [comment]
Okay, will do.

I just want you to confirm about the plans for using this codebase, this is essentially a key exchange server. We’ve been using gun server for doing that till now. Would this server be replacing gun or would work along with it?

as mentioned here
https://github.com/DimensionDev/Maskbook/issues/2363#issuecomment-842136214

**Comment by Jack-Works** [comment]
> I just want you to confirm about the plans for using this codebase, this is essentially a key exchange server. We’ve been using gun server for doing that till now. Would this server be replacing gun or would work along with it?

Maybe both.

**Comment by iamzubin** [comment]
https://github.com/thisispalash/bounty-maskbook/pull/2

Here’s the PR @Jack-Works

**Comment by iamzubin** [comment]
@Jack-Works @yisiliu

**Comment by Jack-Works** [comment]
Hi, I have reviewed your front-end work (https://github.com/iamzubin/Maskbook/tree/unlock). I found the functionality is basically missing (it only contains a GraphQL query to the Unlock API which the address is hardcoded). I think we should complete the unlock protocol integration because that's the final target of this bounty. That involves a lot of work in the front-end.
The mechanism I described above is used for key distributing in a decentralized way, it's still useful for us in many ways (device sync, group share, ....). However, we are not focusing on this backend service right now and it's bounded by the completion of the frontend interface. Can you tell us your plan on completing this bounty, please?

**Comment by iamzubin** [comment]
The unlock protocol integration works on the GraphQL APIs, Currently the api endpoint is hardcoded (I saw the gun api endpoint hardcoded figured might as well do same for the unlock graph ), That's pretty much it as for the user side unlock protocol integration.

while making the post a unlock nft token address is stored in the localdata [this commit](https://github.com/iamzubin/Maskbook/pull/1/commits/6923721f10032da9a3fe3241e57faa1ccd9da53f) . which would be used later to verify the intended recipients.

when the user encounters the post they would require the key to read it. which would be provided by the original poster or someone else who already has the access to the key, before sending the key to user, provider would verify the request by verifying if account making the request actually has access to the unlock using the function [verifyHolder](https://github.com/iamzubin/Maskbook/pull/1/commits/33f990e03769ae61cb7b5560313f1d3a54136aa6#diff-265b26c58053c4f6cdfea61e29c4f65dbfcbb0d10c172baa1567d05d5226f40cR32) .

is there anything I'm missing on the frontend that needs to be added?

**Comment by iamzubin** [comment]
If nothing I would like to you to take a look at the backend service as it currently does not meet all the requirements that were mentioned in the issue #2520 , before I go ahead writing and adding to the server code I would like to be sure about if that meets the requirements, or are there any major changes to be made.

**Comment by Jack-Works** [comment]
About the front-end part, yes, there're lots of things that need to be changed. It should follow other plugins way:

- Define a new plugin
- Provide an entry for composing new content under the protection of the Unlock protocol (and upload the key to the unlock server)
- Insert the encrypted content and payment address to the metadata section (like other plugins do)
- If a post with Unlock metadata found, it should render the unlock UI and retrieve the key from the unlock API.

**Comment by iamzubin** [comment]
Umm okay this was being discussed on the discord server to integrate unlock as part of the system instead of a plugin.
as to read the metadata we need to decipher the post first, and then user can simply read the data. or we need to encrypt the data twice and would need to perform exchange twice.
![image](https://user-images.githubusercontent.com/8392856/120463677-c962e600-c3b9-11eb-9e30-ef9d22d70e8f.png)
also @yisiliu said it's okay that way.

**Comment by iamzubin** [comment]
I can add a UI to buy the access to the post.
this would require user to know what's the address of the given locks is on the eth chain.

The address can be added on the Gun2 and then user can query the database for the lock address.

OR

We can alter the post data and add the unlock locks on the payload itself.
then user can buy access to the given unlock protocol lock and then request for the keys on the key sharing server.

**Comment by iamzubin** [comment]
OH, the unlock protocol does not provide a key sharing platform.
you have to manage the keys yourself if you like to use the unlock protocol (the key sharing service is not yet released)

**Comment by iamzubin** [comment]
@Jack-Works ?

**Comment by Jack-Works** [comment]
> Umm okay this was being discussed on the discord server to integrate unlock as part of the system instead of a plugin.
as to read the metadata we need to decipher the post first, and then user can simply read the data. or we need to encrypt the data twice and would need to perform exchange twice.

Oh yes, so you can add it in the recipient part. But it's detail should in the metadata section, otherwise it required to integrate this into the payload format. I think we should make the outer post public encrypted and put the encrypted information in the metadata section.

> this would require user to know what's the address of the given locks is on the eth chain.

Yeah this information should be in the metadata too. 

> We can alter the post data and add the unlock locks on the payload itself.

That introduces much more complexity.

> OH, the unlock protocol does not provide a key sharing platform.
you have to manage the keys yourself if you like to use the unlock protocol (the key sharing service is not yet released)

Wait... What? So what does the unlock API provides, exactally? It's only checking if someone pays another one??

**Comment by iamzubin** [comment]
More or less yeah(currently ), so I can’t share the key using unlock protocol.


I know it adds to the complexity, I’m willing to implement it, can we discuss the functionality properly again?

Having conversations one message per day is too slow, we’ve already delayed this and I would like to speed it up.
@Jack-Works

**Comment by Jack-Works** [comment]
> More or less yeah(currently ), so I can’t share the key using unlock protocol.

Ok, I go to the Unlock website and cleared my misunderstanding. I need some time to rethink this.

**Comment by iamzubin** [comment]
> Hey, @yisiliu @Tedko I’ve been trying to contact you guys on telegram but @yisiliu hasn’t been replying since a few weeks and @Tedko stopped replying as well :(
> 
> Can I please get a split of the payment as discussed in the issue here
> [#2363 (comment)](https://github.com/DimensionDev/Maskbook/issues/2363#issuecomment-794577313)
Can I please get a followup on the situation.
there seems to be a blocker out of my scope of work for now, I've done the work depending on the discussion done in the discord channel. I've been trying to contact you guys on telegram but I don't get a reply for long times.

Timeline is very much dependent on when the blocker gets resolved, if that requires me to get involved in I would be happy to.

as discussed with @yisiliu , ( after talking with @Tedko ) last week waiting on @jack-works ' review, I didn't hear back from him after trying to contact him multiple times on telegram, can we please speed up the process.


thanks.

**Comment by iamzubin** [comment]
@yisiliu @Tedko ^

**Comment by iamzubin** [comment]
https://github.com/iamzubin/maskbookUnlock/pull/1

Here's the code for server, it's WIP, but please take a look and comment changes.

**Comment by iamzubin** [comment]
update: you can checkout mu fork (branch: unlockProtocolPlugin), I'm writing the cryptographic functions right now, will PR in a couple of days

**Comment by iamzubin** [comment]
Some minor stuff left to fix, but please take a look

**Comment by Tedko** [comment]
Great. Ready to go? @yisiliu

**Comment by iamzubin** [comment]
just realised there's a check missing for the unlock mechanism, i'll be pushing it today itself.

**Comment by iamzubin** [comment]
I did push the commit, (and one more thing I kinda missed) please take a look and let me know about the changes.
I'm working on cleaning up the server code now.

**Comment by iamzubin** [comment]
hey @julien51 is there a permalink based page for buying unlock locks? so we can directly link it to that page as adding a buy button for each lock will make the ui bloated 
i did make a makeshift one but it's pretty janky.

**Comment by Tedko** [comment]
Hi @https://github.com/julien51 @julien51 can u reply to the bounty hacker?

**Comment by Tedko** [comment]
@iamzubin any update?

**Comment by iamzubin** [comment]
yeah it's done, i was waiting on a way to let users buy locks without bloating the ui

i will use my solution for now and do a pr on the main repository today then.

**Comment by julien51** [comment]
i am so sorry for the late response here! (I suck at Github notifications!)

> hey @julien51 is there a permalink based page for buying unlock locks? so we can directly link it to that page as adding a buy button for each lock will make the ui bloated
i did make a makeshift one but it's pretty janky.

yes, absolutely. See te purchase address:
![image](https://user-images.githubusercontent.com/17735/127123668-6942970c-97b0-4cc4-bbbf-9c6de22da18a.png)


To build that URL, just start with `https://app.unlock-protocol.com/checkout?` and then add a query param for `paywallConfig` which includes an URL encoded config as [explained on this page](https://docs.unlock-protocol.com/developers/paywall/locking-page).

We will actually be adding redirect params as well which lets you redirect the user when they close the page or when they complete the purchase. Please bear with us!

**Comment by iamzubin** [comment]
Ahh okay, till then I'll make the application to work without the redirect url (user will have to reload the page.)

I could not find it in the page, but I'll checkout the source and figure something out. (I'm avoiding including the paywall js as it might make the application a little heavy)

thanks for the reply :)

**Comment by julien51** [comment]
Of course! Sorry for the delay. Please do not hesitate to [reach out on Discord](https://discord.gg/Ah6ZEJyTDp)

**Comment by iamzubin** [comment]
Hey, before making a PR I was trying to run the code in the develop branch, I can't seem to get it to work on my machine.
release branch works fine tho.
should I do a pr to develop branch anyway?

**Comment by iamzubin** [comment]
Ahh okay so how the extensions work changed since when I last updated my repo
I need to make those changes, I'm working on it

**Comment by iamzubin** [comment]
I can't seem to understand the conversion, can you please tell me where to look so I can convert to the new standard of plugins.
meanwhile you can take a look [here](https://github.com/iamzubin/Maskbook/pull/2) as I posted earlier also.

**Comment by Jack-Works** [comment]
> Hey, before making a PR I was trying to run the code in the develop branch, I can't seem to get it to work on my machine.
> release branch works fine tho.
> should I do a pr to develop branch anyway?

What happened?

**Comment by iamzubin** [comment]
working of plugins changed, injected components i wrote are not working in the SNEAdaptor for some reason

**Comment by Jack-Works** [comment]
> working of plugins changed, injected components i wrote are not working in the SNEAdaptor for some reason

Did you convert your plugin into the new plugin infra and registered it to enable it?

**Comment by iamzubin** [comment]
> > working of plugins changed, injected components i wrote are not working in the SNEAdaptor for some reason
> 
> Did you convert your plugin into the new plugin infra and registered it to enable it?

Yeah, so I'm working on it but am running into endless errors.
right now I'm stuck on ```DecryptedInspector```. but if I'm stuck for longer time I'll let you know

**Comment by Jack-Works** [comment]
FYI here is an example of migrating an plugin: https://github.com/DimensionDev/Maskbook/pull/3430

**Comment by iamzubin** [comment]
Where are we planning to host the server?
Do I need to set it up?

**Comment by yisiliu** [comment]
@iamzubin yeah please seti it up for the demo and once we merge the pr we will do it ourselves.

**Comment by iamzubin** [comment]
Hey, I was not able to set up the server today, for now you can clone the server and run it locally.
here is the [repository ](https://github.com/iamzubin/maskbookUnlock/tree/server)

You can compile the extension code from the one from PR


I will try to host the server tomorrow. till then if you can take a look without me hosting it please do.

**Comment by iamzubin** [comment]
Alright, so I've hosted the code on my server, and you can test the plugin on this branch `unlockProtocoltest`, on my fork.
[link](https://github.com/iamzubin/Maskbook/tree/unlockProtocoltest)
Please test and lemme know if you run into any issues.

**Comment by iamzubin** [comment]
You can checkout this [tweet](https://twitter.com/primeredditor/status/1424111470652973060) (other tweets from the same account won't work as I've not migrated the local database), You'll need to buy a lock on rinkeby test network to look at the tweet.

**Comment by iamzubin** [comment]
@yisiliu @Jack-Works @Tedko

**Comment by iamzubin** [comment]
hey @yisiliu any updates?

**Comment by Jack-Works** [comment]
> Alright, so I've hosted the code on my server, and you can test the plugin on this branch `unlockProtocoltest`, on my fork.
> [link](https://github.com/iamzubin/Maskbook/tree/unlockProtocoltest)
> Please test and lemme know if you run into any issues.

Can you commit your server URL as the test server so we don't have to set up another one?

**Comment by iamzubin** [comment]
it's on the branch I've linked to

**Comment by iamzubin** [comment]
```export const keyServerEndpoint = 'https://unlock.iamzub.in'```

**Comment by julien51** [comment]
Very excited to see all the progress here!

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. **Started** 3. Submitted 4. Done 

<hr>

__Work for 6000.0 DAI (6000.0 USD @ $1.0/DAI) has been submitted by__:

<hr>

* Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701)
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/DimensionDev/Maskbook/2363/100024701).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://discord.gg/gitcoin/'>Gitcoin's Discord</a>
* $4,314,400.36 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)
